### ISSUE TITLE
C reference implementation: #defines in blake.h should either be tunable, a required part of the interface, or moved to blake3_impl.h

### ISSUE THREAD

**Issue by larryhastings** [BLAKE3-team/BLAKE3#151]
blake3.h has this paragraph of defines:

`    #define BLAKE3_KEY_LEN 32
    #define BLAKE3_OUT_LEN 32
    #define BLAKE3_BLOCK_LEN 64
    #define BLAKE3_CHUNK_LEN 1024
    #define BLAKE3_MAX_DEPTH 54
    #define BLAKE3_MAX_SIMD_DEGREE 16`

If any of these are intended to be tunable (using e.g. gcc -D) they should be wrapped in #ifndef:

`    #ifndef BLAKE3_MAX_SIMD_DEGREE
    #define BLAKE3_MAX_SIMD_DEGREE 16
    #endif /* BLAKE3_MAX_SIMD_DEGREE */`

If they're not tunable, then they should probably be moved to blake3_impl.h.  The exception is if they're used as part of the external API, e.g. `BLAKE3_BLOCK_LEN` is used in the definition of the `blake3_chunk_state` struct.

**Comment by sneves** [comment]
They're not tunable. `BLAKE3_OUT_LEN`, `BLAKE3_BLOCK_LEN`, `BLAKE3_MAX_DEPTH` are directly necessary in the definitions below.

The other values are not used, but are useful to know. Arguably, all of those ought to be defined as an enumeration instead of macros.

**Comment by oconnor663** [comment]
Yeah definitely not tunable but sometimes useful. For example, using CHUNK_LEN * MAX_SIMD_DEGREE can make sense for a static buffer size, if you want to guarantee you get all the SIMD optimizations you can possibly get. (I did just remove `BLAKE3_MAX_SIMD_DEGREE` as part of https://github.com/BLAKE3-team/BLAKE3/issues/152, partly because it was duplicative and also kind of incorrect, but that could be an argument for adding it back.)

What's the effective difference between an enum and a preprocessor definition with these things?

**Comment by larryhastings** [comment]
> What's the effective difference between an enum and a preprocessor definition with these things?

Not a lot.  A preprocessor definition can be redefined with a subsequent `#undef` and `#define`.  Also, you can set preprocessor defines from the command-line (gcc -D), which you can't do with enums.

An enum can't be changed, though it _can_ be papered over with a preprocessor `#define`, as the preprocessor definition would take precedence.

I'm not sure why one would prefer an enum over a `const int` here.  blake3's header file already uses `const`, so it's not like backwards compatibility with ancient dialects of C is a concern.

**Comment by sneves** [comment]
A `const int` would have an address, occupy space, and would thereby be involved in linking, and would need a split definition between header and source, for it to be included in different files linked together.  A `static const int` would not cause linking problems, but would result in wasted space as well. There's no reason for compile-time constants to need storage, or an address.

**Comment by larryhastings** [comment]
I'm pretty sure the compiler will optimize away a "static const int" so it doesn't consume any storage / does not have an address, if you never take its address.

Anyway, if it were me, I'd define untunable constants the way you already have, as unconditional preprocessor macros.

**Comment by wkarshat** [comment]
A quick check of the sample code in main.c and example.c indicates that KEY_LEN and OUT_LEN are very much a part of the interface [standard] and thus belong in blake3.h
The file already has #ifndef BLAKE3_H wrapper
Using #define for an integer constant is so canonical in straight C, compared with the enum usage, every embedded engineer will get it.
 
BLOCK_LEN and CHUNK_LEN are only used in the implementation files, not sample code, and likely belong in the _impl.h What use are these, and why are they not tunable, at least in principle?

The two BLAKE3_MAX defines indeed are not used, so what's the point having them exactly?

**Comment by oconnor663** [comment]
> BLOCK_LEN and CHUNK_LEN are only used in the implementation files, not sample code, and likely belong in the _impl.h What use are these, and why are they not tunable, at least in principle?

The places where I've wanted them in the past are for sizing buffers, and to a lesser extent for benchmarks and test vectors. Typically you want to buffer a power-of-2 number of chunks when you're hashing some stream incrementally.

If you were to _change_ the value of those constants, at best you'd have an incompatible hash function, with different output. The reference implementation might even allow this, if all you do is change the chunk size to some other power of two. But in practice in the production implementation, the value of these constants is assumed and hardcoded in a variety of ways throughout the SIMD implementations, so what would actually happen is you'd either start crashing or start reading/writing uninitialized memory in some awful way 😬

**Comment by larryhastings** [comment]
Well, my question is, do external people calling into the BLAKE3 API need these constants?  In the case of `BLAKE3_BLOCK_LEN`, yes, because it's used in the definition of `blake3_chunk_state`.  But  `BLAKE3_CHUNK_LEN` isn't part of the API.  Certainly it's an important implementation detail.  But external users don't need it and aren't allowed to tune it.  So I suggest it should move to `blake3_impl.h`.  Unless I've missed something, and it's important for external callers into BLAKE3 to understand this number for some reason.

**Comment by oconnor663** [comment]
Ah I guess the unstated background here is that I've imagined `blake3_impl.h` is a private header, which shouldn't be included by calling code and isn't guaranteed to be stable. Should probably have put a comment in there to that effect... What do other C projects do around public/private conventions?

**Comment by larryhastings** [comment]
I assumed `blake3_impl.h` was a private header file based on the name.

As for, what other projects do, approaches differ.  Just in the last few years Python created an "Include/internal" directory and moved all the private stuff into there.  If I worked on BLAKE3, I'd probably put all the public header files in an "include" directory, and move the C files and private header files into a "src" directory, which I assert is pretty common.
